File: src/lexer/lexer.ts
import { Location } from "../diagnostic/location";
import { Logger } from "../diagnostic/logger";
import { Result } from "../lib/result";
import { Token, TokenType, type TokenTypeData } from "./token";

export class Lexer {
    private static tokenTypes
        = Object.entries(TokenType) as [keyof typeof TokenType, TokenTypeData][];

    constructor(
        public input: string,
        public location: Location = new Location(0, 1, 1)
    ) {}

    public lex(logger: Logger): Result<Token[]> {
        const tokens = [];

        while (this.location.index < this.input.length) {
            let result = this.getNextToken(logger);

            if (result)
                tokens.push(result);
            else {
                while (this.location.index < this.input.length && (result = this.getNextToken(logger)) === null)
                    this.advance();

                if (result)
                    tokens.push(result);
            }
        }

        if (logger.diagnostics.length > 0)
            return Result.failure(new Error());

        tokens.push(new Token("EndOfInput", "", this.location.clone(), null));
        return Result.success(tokens);
    }

    public getNextToken(logger: Logger): Token<any> | null {
        const start = this.location.clone();
        const input = this.input.substring(start.index);
        const character = input[0];

        // console.log("Getting new token", input);

        if (character.trim() === "") {
            const length = input.length - input.trimStart().length;
            this.advance(length);
            return new Token("Whitespace", input.substring(0, length), start, null);
        }

        if (Lexer.isLetter(character)) {
            let identifier = character;
            let char: string;

            while (Lexer.isLetterOrDigit(char = this.input.charAt(this.advance() + 1)))
                identifier += char;

            if (identifier === "true" || identifier === "false")
                return new Token<boolean>("Boolean", identifier, start, identifier === "true");

            for (const [key, value] of Lexer.tokenTypes)
                if (value.value === identifier)
                    return new Token(key, identifier, start, null);

            return new Token("Identifier", identifier, start, null);
        }

        if (Lexer.isDigit(character)) {
            let hasSeparator = false;
            let multiplier = 0.1;
            let raw = character;
            let value = this.digitToInt(character);
            let char;

            while (Lexer.isDigit(char = this.input.charAt(this.advance() + 1)) || (!hasSeparator && char === ".")) {
                if (char === ".") {
                    hasSeparator = true;
                    continue;
                }

                const digitValue = this.digitToInt(char);

                if (hasSeparator) {
                    value += digitValue * multiplier;
                    multiplier /= 10;
                } else {
                    value *= 10;
                    value += digitValue;
                }

                raw += char;
            }

            return new Token<number>("Number", raw, start, value);
        }

        if (character === "\"") {
            let raw = character;
            let char: string;
            let value = "";
            this.advance(1);

            while ((char = this.input.charAt(this.advance())) !== "\"") {
                raw += char;

                if (char === "\\") {
                    const next = this.input.charAt(this.advance());
                    raw += next;
                    value += next;
                } else {
                    value += char;
                }
            }

            return new Token<string>("String", raw, start, value);
        }

        for (const [key, value] of Lexer.tokenTypes)
            if (value.value && input.startsWith(value.value))
                return new Token(key, value.value, (this.advance(value.value.length), start), null);

        logger.error(`Unexpected character '${character}'`, this.location.clone());
        return null;
    }

    public static isLetter(character: string): boolean {
        const a = 97;
        const z = 122;
        const A = 65;
        const Z = 90;
        const code = character.charCodeAt(0);
        return (code >= a && code <= z) || (code >= A && code <= Z);
    }

    public static isDigit(character: string): boolean {
        const zero = 48;
        const nine = 57;
        const code = character.charCodeAt(0);
        return code >= zero && code <= nine;
    }

    public static isLetterOrDigit(character: string): boolean {
        return Lexer.isLetter(character) || Lexer.isDigit(character);
    }

    private digitToInt(character: string): number {
        const zero = 48;
        const code = character.charCodeAt(0);
        return code - zero;
    }

    private advance(amount: number = 1): number {
        const oldIndex = this.location.index;
        const advancement = this.input.substring(this.location.index, this.location.index + amount);
        this.location.index += amount;
        
        for (const char of advancement)
            if (char === "\n") {
                this.location.line++;
                this.location.column = 1;
            } else this.location.column++;

        return oldIndex;
    }
}

File: src/main/kotlin/me/honkling/miniscript/lexer/Lexer.kt
package me.honkling.miniscript.lexer

import me.honkling.miniscript.diagnostic.Location
import me.honkling.miniscript.diagnostic.Logger
import me.honkling.miniscript.diagnostic.MiniScriptException

private const val newLine = '\n'
private const val escape = '\\'
private const val doubleQuotes = '"'
private const val singleQuotes = '\''
private const val period = '.'

class Lexer(
    private val input: String,
    private val logger: Logger,
    private val location: Location = logger.location,
    private val internalMode: Boolean = false
) {
    fun lex(): List<Token> {
        val tokens = mutableListOf<Token>()

        while (!isEnd())
            try {
                tokens += lexToken()
            } catch (_: MiniScriptException.LexError) {}

        tokens += Token(TokenType.EndOfInput, "", location.clone(), logger)
        return tokens
    }

    private fun lexToken(): Token {
        val character = input[location.index]

        if (character == singleQuotes)
            return lexCharacter()

        if (character == doubleQuotes)
            return lexString()

        if (character.isDigit())
            return lexNumber()

        if (character.isWhitespace())
            return lexGreedyTokenByPredicate(TokenType.Whitespace, Char::isWhitespace)

        if (character.isLetter() || character == '_') {
            val start = location.clone()
            val raw = lexGreedyStringByPredicate { char, index ->
                (index == 0 && (char.isLetter() || char == '_')) || char.isLetterOrDigit() || char == '_'
            }

            val isTrue = raw == "true"
            if (isTrue || raw == "false")
                return Token.WithValue(TokenType.Boolean, raw, start, logger, isTrue)

            for (type in TokenType.entries)
                if (type.value == raw)
                    return Token(type, raw, start, logger)

            return Token(TokenType.Identifier, raw, start, logger)
        }

        for (type in TokenType.entries)
            if (input.startsWith(type.value ?: continue, location.index))
                return Token(
                    if (type == TokenType.Native && !internalMode)
                        TokenType.Identifier
                    else type,
                    type.value,
                    advance(type.value.length),
                    logger
                )

        advance(1)
        logger.error("Unexpected character found during lexing")
        throw MiniScriptException.LexError()
    }

    private fun lexCharacter(): Token.WithValue<Char> {
        val start = location.clone()
        val raw = StringBuilder()

        raw.append(input[advance(1).index])
        val consumedCharacter = input.getOrNull(advance(1).index)
        raw.append(consumedCharacter)

        val character = when (consumedCharacter) {
            escape -> {
                val character = input.getOrNull(advance(1).index)
                raw.append(character)
                character
            }
            else -> consumedCharacter
        }

        if (character == null) {
            logger.error("Expected a character, found nothing")
            throw MiniScriptException.LexError()
        }

        val endingCharacter = input.getOrNull(advance(1).index)
        if (endingCharacter != singleQuotes) {
            logger.error("Expected \"'\", found ${endingCharacter?.let { "'$it'" } ?: "nothing"}")
            throw MiniScriptException.LexError()
        }

        raw.append(endingCharacter)
        return Token.WithValue(TokenType.Character, raw.toString(), start, logger, character)
    }

    private fun lexNumber(): Token.WithValue<Double> {
        val start = location.clone()
        val raw = StringBuilder()
        var hasPeriod = false
        var divisor = 0.1
        var value = 0.0

        while (!isEnd() && (input[location.index].isDigit() || (!hasPeriod && input[location.index] == period))) {
            val character = input[location.index++]

            raw.append(character)

            if (character == period) {
                hasPeriod = true
                continue
            }

            val digitValue = character.digitToInt()

            if (!hasPeriod) {
                value *= 10
                value += digitValue
            } else {
                value += digitValue * divisor
                divisor /= 10
            }
        }

        return Token.WithValue(TokenType.Number, raw.toString(), start, logger, value)
    }

    private fun lexString(): Token.WithValue<String> {
        val start = location.clone()
        val value = StringBuilder()
        val raw = StringBuilder("\"")
        advance(1)

        while (true) {
            val character = input.getOrNull(advance(1).index)

            if (isEnd() || character == newLine) {
                val representation = "end of " + (character?.let { "line" } ?: "file")
                logger.error("Expected string to close, found $representation")
                throw MiniScriptException.LexError()
            }

            raw.append(character)

            when (character) {
                doubleQuotes -> break
                escape -> {
                    val escapedCharacter = input.getOrNull(advance(1).index)

                    if (escapedCharacter == null) {
                        logger.error("Expected character to escape, found end of file")
                        throw MiniScriptException.LexError()
                    }

                    raw.append(escapedCharacter)

                    if (escapedCharacter == 'n')
                        value.append("\n")
                    else if (escapedCharacter == 't')
                        value.append("\t")
                    else value.append(escapedCharacter)
                }
                else -> value.append(character)
            }
        }

        return Token.WithValue(TokenType.String, raw.toString(), start, logger, value.toString())
    }

    private fun lexGreedyTokenByPredicate(type: TokenType, predicate: (Char) -> Boolean)
        = lexGreedyTokenByPredicate(type) { char, _ -> predicate(char) }

    private fun lexGreedyTokenByPredicate(type: TokenType, predicate: (Char, Int) -> Boolean): Token {
        val start = location.clone()
        return Token(type, lexGreedyStringByPredicate(predicate), start, logger)
    }

    private fun lexGreedyStringByPredicate(predicate: (Char) -> Boolean)
            = lexGreedyStringByPredicate { char, _ -> predicate(char) }

    private fun lexGreedyStringByPredicate(predicate: (Char, Int) -> Boolean): String {
        val raw = StringBuilder()

        while (!isEnd() && predicate(input[location.index], location.index))
            raw.append(input[advance(1).index])

        return raw.toString()
    }
    
    private fun advance(length: Int): Location {
        val start = location.clone()
        
        repeat(length) {
            val character = input[location.index++]
            
            if (character == '\n') {
                location.line++
                location.column = 1
            } else location.column++
        }
        
        return start
    }

    private fun isEnd()
        = location.index >= input.length
}